File: android/app/src/main/java/app/organicmaps/widget/RoutingToolbarButton.java
package app.organicmaps.widget;

import android.content.Context;
import android.util.AttributeSet;
import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.AppCompatRadioButton;
import app.organicmaps.R;
import app.organicmaps.util.ThemeUtils;

public class RoutingToolbarButton extends AppCompatRadioButton
{
  private boolean mInProgress;
  @DrawableRes
  private int mIcon;

  public RoutingToolbarButton(Context context, AttributeSet attrs, int defStyleAttr)
  {
    super(context, attrs, defStyleAttr);
    initView();
  }

  public RoutingToolbarButton(Context context, AttributeSet attrs)
  {
    super(context, attrs);
    initView();
  }

  public RoutingToolbarButton(Context context)
  {
    super(context);
    initView();
  }

  private void initView()
  {
    final boolean isNightTheme = ThemeUtils.isNightTheme();
    setBackgroundResource(isNightTheme ? R.drawable.routing_toolbar_button_night : R.drawable.routing_toolbar_button);
    setButtonTintList(isNightTheme ? R.color.routing_toolbar_icon_tint_night : R.color.routing_toolbar_icon_tint);
  }

  public void progress()
  {
    if (mInProgress)
      return;

    setButtonDrawable(mIcon);
    mInProgress = true;
    setActivated(false);
    setSelected(true);
  }

  public void activate()
  {
    if (!mInProgress)
    {
      setButtonDrawable(mIcon);
      setSelected(false);
      setActivated(true);
    }
  }

  public void complete()
  {
    mInProgress = false;
    activate();
  }

  public void deactivate()
  {
    setActivated(false);
    mInProgress = false;
  }

  public void setButtonTintList(@ColorRes int color)
  {
    setButtonTintList(AppCompatResources.getColorStateList(getContext(), color));
  }

  public void setIcon(@DrawableRes int icon)
  {
    mIcon = icon;
    setButtonDrawable(icon);
  }
}

File: android/app/src/main/java/app/organicmaps/widget/RoutingToolbarButton.kt
package app.organicmaps.widget

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.view.Gravity
import androidx.appcompat.widget.AppCompatRadioButton
import androidx.core.content.ContextCompat
import androidx.core.widget.CompoundButtonCompat
import app.organicmaps.R

class RoutingToolbarButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : AppCompatRadioButton(context, attrs, defStyleAttr) {
    init {
        setBackgroundResource(R.drawable.routing_toolbar_button)
        buttonTintList = ContextCompat.getColorStateList(context, R.color.routing_toolbar_icon_tint)
        setPadding(0, 0, 0, 0)
        compoundDrawablePadding = 0
        gravity = Gravity.CENTER
    }

    // RadioButton draws the button drawable at the start edge; on a stretched (weight=1) cell
    // inside the routing pill, translate horizontally so the icon stays centered.
    override fun onDraw(canvas: Canvas) {
        val drawable = CompoundButtonCompat.getButtonDrawable(this) ?: return super.onDraw(canvas)
        val shift = (width - drawable.intrinsicWidth) / 2
        if (shift <= 0) return super.onDraw(canvas)
        val dx = if (layoutDirection == LAYOUT_DIRECTION_RTL) -shift else shift
        canvas.save()
        canvas.translate(dx.toFloat(), 0f)
        super.onDraw(canvas)
        canvas.restore()
    }
}
